### PR TITLE
Refactor: Creating entity classes and populating them seaparately to reuse them between reports

### PR DESCRIPTION
**Pull request by aanil**
* Created objects Project, Sample, Prep, Flowcell, Lane and AbortedSampleInfo to hold data that can be used by different report types. 
* Refactored project summary report code to use the above objects to generate the report.
* Fixed a bug for wrong option selected for NextSeq in lane cluster reads.
* Removed piper xml parsing from ngi_reports as piper has been replaced by sarek.

### CODE REVIEW

**Review by ssjunnebo — APPROVED** [review]
Awesome work! Looks good to me, but it's a massive PR so I might have missed something. Let's see what the testing uncovers 😉

**Comment by sofiahaglund** [comment]
I can't comment that much on the code, but if it looks good to Sara, it looks good to me :-)

I have tested it now, and it all looks good, good job!

**Review by sofiahaglund — APPROVED** [review]

File: ngi_reports/utils/entities.py [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+        self.report_date = ''
+        self.sequencing_setup = ''
+        self.skip_fastq = False
+        self.user_ID = ''
+
+    def populate(self, log, organism_names, **kwargs):
+
+        project = kwargs.get('project', '')
+        if not project:
+            raise SystemExit()
```
Comment (alneberg) [inline ngi_reports/utils/entities.py:135]: This causes a silent exit which might not even indicate an error. If this is a true error, raise an exception with a message instead.

File: ngi_reports/utils/entities.py [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+
+        if proj.get('source') != 'lims':
+            log.error('The source for data for project {} is not LIMS.'.format(project))
+            raise BaseException
+
+        proj_details = proj.get('details',{})
+
+        if 'aborted' in proj_details:
+            log.warn('Project {} was aborted, so not proceeding.'.format(project))
+            raise SystemExit
```
Comment (alneberg) [inline ngi_reports/utils/entities.py:164]: See above, I think we should get rid of the silent SystemExit

File: ngi_reports/utils/entities.py [hunk 3]
```diff
@@ -0,0 +1,453 @@
...
+                        r_num = len(r_len_list)
+                        qval = float(stat.get(qval_key))
+                        pfrd = int(stat.get(base_key).replace(',',''))
+                        pfrd = pfrd/2 if fc['db'] == 'flowcell' else pfrd
+                        base = pfrd * sum(r_len_list)
+                        sample_qval[sample][r_idx] = {'qval': qval, 'reads': pfrd, 'bases': base}
+
+                    except (TypeError, ValueError, AttributeError) as e:
+                        log.warn('Something went wrong while fetching Q30 for sample {} with barcode {} in FC {} at lane {}'.format(sample, barcode, fcObj.name, lane))
+                        pass
```
Comment (alneberg) [inline ngi_reports/utils/entities.py:373]: I'm not a fan of this general try-except-pass section. I guess we don't really know what's going wrong and it seems like we don't care?

File: ngi_reports/utils/entities.py [hunk 4]
```diff
@@ -0,0 +1,453 @@
...
+                        if kwargs.get('fc_phix',{}).get(fcObj.name, {}):
+                            laneObj.phix = kwargs.get('fc_phix').get(fcObj.name).get(lane)
+
+                        fcObj.lanes[lane] = laneObj
+
+                        ## Check if the above created lane object has all needed info
+                        for k,v in vars(laneObj).items():
+                            if not v:
+                                log.warn('Could not fetch {} for FC {} at lane {}'.format(k, fcObj.name, lane))
+                except KeyError:
```
Comment (alneberg) [inline ngi_reports/utils/entities.py:393]: GAH! Any KeyError within the previous 40 lines of code is just fine? Super easy to miss some more serious errors this way.

Comment (alneberg) [reply ngi_reports/utils/entities.py:393]: But I guess it just affects the quality values anyway...

**Review by alneberg — APPROVED** [review]
Looks good! I'd like to see the project argument moved to the constructor for the report. I guess that will result in a lot of replacing `project` with `self.project` but that should be simple enough? Also couldn't resist commenting on the large try except clauses even though I realise their not your doings.

File: ngi_reports/ngi_reports.py [hunk 5]
```diff
@@ -33,8 +34,11 @@ def make_reports (report_type, working_dir=os.getcwd(), config_file=None, **kwar
     # Import the modules for this report type
     report_mod = __import__('ngi_reports.reports.{}'.format(report_type), fromlist=['ngi_reports.reports'])
 
+    proj = Project()
+    proj.populate(LOG, config._sections['organism_names'], **kwargs)
+
     # Make the report object
-    report = report_mod.Report(config, LOG, working_dir, **kwargs)
+    report = report_mod.Report(LOG, working_dir, **kwargs)
```
[the remark this answers is not in the document]
Comment (aanil) [reply ngi_reports/ngi_reports.py:41]: Agree to disagree for now?

File: ngi_reports/utils/entities.py (hunk 1 above)
Comment (aanil) [reply ngi_reports/utils/entities.py]: Done!

File: ngi_reports/utils/entities.py (hunk 2 above)
Comment (aanil) [reply ngi_reports/utils/entities.py]: Done!

File: ngi_reports/utils/entities.py (hunk 4 above)
Comment (aanil) [reply ngi_reports/utils/entities.py]: Changed it somewhat based on what I think it was originally for.

**Comment by alneberg** [comment]
Very nice work!

**Pull request merged**